<a href="https://colab.research.google.com/github/mugalan/introduction-to-statistical-learning/blob/main/sample-answers/max_entropy_inference_assignment_answers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Maximum Entropy Inference

# Q1. Statistical Mechanics of the Quantum Oscillator

## 1.1 The Partition Function

The energy levels of a single quantum harmonic oscillator of angular frequency $\omega$ are

$$E_j = \hbar\omega\left(j+\tfrac12\right), \qquad j = 0,1,2,\dots$$

so with $\beta \triangleq 1/(k_BT)$ the canonical partition function is

$$Z(T) = \sum_{j=0}^{\infty} e^{-\beta E_j} = \sum_{j=0}^\infty e^{-\beta\hbar\omega(j+1/2)} = e^{-\beta\hbar\omega/2}\sum_{j=0}^\infty \left(e^{-\beta\hbar\omega}\right)^{j}.$$

Since $T>0 \Rightarrow \beta\hbar\omega>0 \Rightarrow 0<e^{-\beta\hbar\omega}<1$, the sum is a **convergent geometric series** with ratio $r=e^{-\beta\hbar\omega}$, so $\sum_{j=0}^\infty r^j = 1/(1-r)$. Hence

$$\boxed{Z(T) = \frac{e^{-\hbar\omega/(2k_BT)}}{1-e^{-\hbar\omega/(k_BT)}}}$$

It is convenient to write $x \triangleq \dfrac{\hbar\omega}{k_BT}$, so $Z = \dfrac{e^{-x/2}}{1-e^{-x}}$ and $\ln Z = -\dfrac{x}{2} - \ln\left(1-e^{-x}\right)$.

## 1.2 Thermodynamic Potentials

**Helmholtz Free Energy.**

$$F(T) = -k_BT\ln Z = -k_BT\left[-\frac{\hbar\omega}{2k_BT} - \ln\left(1-e^{-\hbar\omega/k_BT}\right)\right]$$

$$\boxed{F(T) = \frac{\hbar\omega}{2} + k_BT\ln\left(1-e^{-\hbar\omega/k_BT}\right)}$$

**Internal Energy.** Using $U=-\dfrac{\partial \ln Z}{\partial \beta}$ with $\ln Z = -\beta\hbar\omega/2-\ln(1-e^{-\beta\hbar\omega})$:

$$\frac{\partial \ln Z}{\partial \beta} = -\frac{\hbar\omega}{2} - \frac{\hbar\omega\, e^{-\beta\hbar\omega}}{1-e^{-\beta\hbar\omega}}
\;\;\Longrightarrow\;\;
U = \frac{\hbar\omega}{2} + \frac{\hbar\omega\, e^{-\beta\hbar\omega}}{1-e^{-\beta\hbar\omega}}$$

$$\boxed{U(T) = \hbar\omega\left(\frac12 + \frac{1}{e^{\hbar\omega/k_BT}-1}\right)}$$

This is the familiar zero-point energy $\hbar\omega/2$ plus $\hbar\omega$ times the Bose–Einstein occupation number $\langle n\rangle = 1/(e^{\hbar\omega/k_BT}-1)$.

## 1.3 Entropy

Using $S = -\left(\dfrac{\partial F}{\partial T}\right)$ with $x=\hbar\omega/(k_BT)$ (so $dx/dT=-x/T$):

$$\frac{\partial F}{\partial T} = k_B\ln(1-e^{-x}) + k_BT\cdot\frac{e^{-x}}{1-e^{-x}}\cdot\left(-\frac{x}{T}\right) = k_B\ln(1-e^{-x}) - \frac{k_B x\, e^{-x}}{1-e^{-x}}$$

so, using $e^{-x}/(1-e^{-x}) = 1/(e^x-1)$,

$$\boxed{S(T) = -k_B\ln\left(1-e^{-\hbar\omega/k_BT}\right) + \frac{\hbar\omega}{T}\cdot\frac{1}{e^{\hbar\omega/k_BT}-1}}$$

**Check that $U = F+TS$:**

$$F+TS = \frac{\hbar\omega}{2} + k_BT\ln(1-e^{-x}) + T\left[-k_B\ln(1-e^{-x}) + \frac{\hbar\omega}{T}\frac{1}{e^x-1}\right]$$

The two $k_BT\ln(1-e^{-x})$ terms cancel exactly, leaving

$$F+TS = \frac{\hbar\omega}{2} + \frac{\hbar\omega}{e^x-1} = U \;\checkmark$$

## 1.4 Heat Capacity and Hyperbolic Form

Differentiating $U(T) = \hbar\omega/2 + \hbar\omega/(e^x-1)$ with respect to $T$ (again $dx/dT=-x/T$):

$$C_V = \frac{\partial U}{\partial T} = \hbar\omega\cdot\left(-\frac{e^x}{(e^x-1)^2}\right)\cdot\left(-\frac{x}{T}\right) = \frac{\hbar\omega\, x}{T}\cdot\frac{e^x}{(e^x-1)^2}$$

Since $x/T = \hbar\omega/(k_BT^2)$,

$$\boxed{C_V = k_B\left(\frac{\hbar\omega}{k_BT}\right)^2 \frac{e^{\hbar\omega/k_BT}}{\left(e^{\hbar\omega/k_BT}-1\right)^2}}$$

**Hyperbolic rewriting.** Factor $e^x-1 = e^{x/2}\left(e^{x/2}-e^{-x/2}\right)$, so

$$(e^x-1)^2 = e^{x}\left(e^{x/2}-e^{-x/2}\right)^2 \;\Longrightarrow\; \frac{e^x}{(e^x-1)^2} = \frac{1}{\left(e^{x/2}-e^{-x/2}\right)^2} = \frac{1}{\left(2\sinh(x/2)\right)^2} = \frac{1}{4\sinh^2(x/2)}$$

Substituting back with $x=\hbar\omega/(k_BT)$, i.e. $x/2 = \hbar\omega/(2k_BT)$:

$$C_V = k_B x^2 \cdot \frac{1}{4\sinh^2(x/2)} = k_B\left(\frac{x}{2}\right)^2\frac{1}{\sinh^2(x/2)}$$

$$\boxed{C_V = k_B\left(\frac{\hbar\omega}{2k_BT}\right)^2\frac{1}{\sinh^2\!\left(\dfrac{\hbar\omega}{2k_BT}\right)}}$$

exactly the requested form.

## 1.5 High-Temperature Limit ($k_BT \gg \hbar\omega$)

Let $x = \hbar\omega/(k_BT)\to 0^+$. Using $\sinh(x/2)\approx x/2$ for small $x/2$:

$$C_V \approx k_B\left(\frac{x}{2}\right)^2\frac{1}{(x/2)^2} = k_B$$

So $C_V \to k_B$ as $T\to\infty$ — **independent of $\omega$ and $\hbar$**. This exactly matches the **classical equipartition theorem**: a 1-D harmonic oscillator has one quadratic kinetic-energy term and one quadratic potential-energy term, each contributing $\tfrac12 k_B$ to the heat capacity, for a total of $k_B$. Quantum effects (the discreteness of $E_j$) only become important, and $C_V$ falls below $k_B$ ("freezes out"), when $k_BT$ is no longer large compared with the level spacing $\hbar\omega$.

In [8]:
import numpy as np
import plotly.graph_objects as go

# Verify Cv both forms agree numerically and show the classical/quantum crossover
hbar_omega_over_kB = 1.0  # set hbar*omega/k_B = 1 (units of temperature)
T = np.linspace(0.02, 5, 500)
x = hbar_omega_over_kB / T

Cv_exp = (x**2) * np.exp(x) / (np.exp(x)-1)**2          # exponential form (in units of k_B)
Cv_sinh = (x/2)**2 / np.sinh(x/2)**2                     # hyperbolic form (in units of k_B)

fig = go.Figure()
fig.add_trace(go.Scatter(x=T, y=Cv_exp, mode='lines', name='C_V (exponential form)', line=dict(width=4)))
fig.add_trace(go.Scatter(x=T, y=Cv_sinh, mode='lines', name='C_V (sinh form)', line=dict(width=2, dash='dash', color='white')))
fig.add_hline(y=1.0, line=dict(color='red', dash='dot'), annotation_text='Classical equipartition limit  C_V = k_B')

fig.update_layout(title="Heat Capacity of the Quantum Harmonic Oscillator",
                   xaxis_title="Temperature T  (units of ħω/k_B)",
                   yaxis_title="C_V / k_B",
                   template="plotly_white", width=800, height=480)
fig.show()


# Q2. The Maximum Entropy Distribution under Mean and Variance Constraints

We seek the density $p(x)$ on $X\in(-\infty,\infty)$ (uniform reference measure $m(x)=1$) that maximizes the differential entropy

$$H[p] = -\int_{-\infty}^{\infty} p(x)\ln p(x)\,dx$$

subject to

$$\int p(x)\,dx = 1, \qquad \int x\,p(x)\,dx = \mu, \qquad \int (x-\mu)^2 p(x)\,dx = \sigma^2.$$

**Step 1 — Lagrangian.** Introduce multipliers $\gamma,\lambda_1,\lambda_2$:

$$\mathscr{L}[p] = -\int p\ln p\,dx - \gamma\left(\int p\,dx - 1\right) - \lambda_1\left(\int xp\,dx-\mu\right) - \lambda_2\left(\int (x-\mu)^2p\,dx-\sigma^2\right)$$

**Step 2 — Euler–Lagrange (functional) derivative.** Setting $\delta\mathscr{L}/\delta p(x)=0$:

$$-\ln p(x) - 1 - \gamma - \lambda_1 x - \lambda_2(x-\mu)^2 = 0$$

$$\Longrightarrow\quad p^*(x) = \exp\left[-(1+\gamma)-\lambda_1 x - \lambda_2(x-\mu)^2\right] = A\,e^{-\lambda_1 x - \lambda_2(x-\mu)^2}, \quad A\triangleq e^{-(1+\gamma)}$$

**Step 3 — Eliminate $\lambda_1$ using the mean constraint.** Substitute $u=x-\mu$ and complete the square:

$$-\lambda_1(u+\mu) - \lambda_2 u^2 = -\lambda_2\left(u+\frac{\lambda_1}{2\lambda_2}\right)^2 + \frac{\lambda_1^2}{4\lambda_2} - \lambda_1\mu$$

so $p^*(x)$ is (up to constants) a Gaussian in $u$ **centered at** $u_0=-\lambda_1/(2\lambda_2)$. For the mean of $X=\mu+u$ to equal $\mu$ we need $\mathbb{E}[u]=0$, i.e. the Gaussian must be centered at $u=0$. This forces

$$\boxed{\lambda_1 = 0}$$

so $p^*(x) = A\,e^{-\lambda_2(x-\mu)^2}$ — a Gaussian centered exactly at $\mu$, as required.

**Step 4 — Normalize using the Gaussian integral** $\displaystyle\int_{-\infty}^\infty e^{-au^2}\,du = \sqrt{\pi/a}$:

$$\int A\,e^{-\lambda_2 u^2}\,du = A\sqrt{\frac{\pi}{\lambda_2}} = 1 \quad\Longrightarrow\quad A = \sqrt{\frac{\lambda_2}{\pi}}$$

**Step 5 — Match the variance constraint using** $\displaystyle\int_{-\infty}^{\infty} u^2 e^{-au^2}\,du = \frac{1}{2a}\sqrt{\frac{\pi}{a}}$:

$$\sigma^2 = \int u^2\,A\,e^{-\lambda_2u^2}\,du = A\cdot\frac{1}{2\lambda_2}\sqrt{\frac{\pi}{\lambda_2}} = \frac{1}{2\lambda_2}$$

$$\Longrightarrow\quad \boxed{\lambda_2 = \frac{1}{2\sigma^2}}, \qquad A = \sqrt{\frac{1}{2\pi\sigma^2}}$$

**Step 6 — Final result.** Substituting $\lambda_1=0,\ \lambda_2=1/(2\sigma^2)$ back:

$$\boxed{p^*(x) = \frac{1}{\sqrt{2\pi\sigma^2}}\exp\left(-\frac{(x-\mu)^2}{2\sigma^2}\right)}$$

which is exactly the Normal density $\mathcal{N}(\mu,\sigma^2)$. Since the entropy functional is strictly concave and the constraint set is affine, this stationary point is the **unique global maximizer**: among *all* distributions on $\mathbb{R}$ with a given mean and variance, the Gaussian is the least-biased (most-uncertain) one — any other choice smuggles in extra, unjustified structure.

In [9]:
import numpy as np
import plotly.graph_objects as go

mu, sigma = 2.0, 1.5
x = np.linspace(mu-5*sigma, mu+5*sigma, 500)
p_gauss = (1/np.sqrt(2*np.pi*sigma**2))*np.exp(-(x-mu)**2/(2*sigma**2))

# Monte-Carlo sanity check: among many random shapes with the SAME mean/variance,
# none should have higher entropy than the Gaussian.
rng = np.random.default_rng(0)
def entropy_hist(samples, bins=200):
    hist, edges = np.histogram(samples, bins=bins, density=True)
    widths = np.diff(edges)
    p = hist[hist>0]
    w = widths[hist>0]
    return -np.sum(p*np.log(p)*w)  # differential entropy estimate

H_gauss = 0.5*np.log(2*np.pi*np.e*sigma**2)  # exact Gaussian differential entropy
print(f"Exact Gaussian differential entropy: {H_gauss:.4f} nats")

fig = go.Figure()
fig.add_trace(go.Scatter(x=x, y=p_gauss, fill='tozeroy', name=f'MaxEnt solution  N({mu},{sigma**2})'))
fig.update_layout(title="Maximum-Entropy Density under Mean & Variance Constraints",
                   xaxis_title="x", yaxis_title="p(x)", template="plotly_white", width=750, height=450)
fig.show()


Exact Gaussian differential entropy: 1.8244 nats


# Q3. Multiclass Logistic Regression: Automatic Classification of Emails

**Modeling note.** The prompt names two labels explicitly, $Y=0$ (Finance) and $Y=1$ (Technical), and *every* empirical constraint given is expressed in terms of $\mathbb{I}(Y=1)$. We therefore treat this as a **binary** MaxEnt / logistic-regression problem (the $C=2$ special case of the softmax result derived in the notes), exactly analogous to the spam-filter example worked in class. The feature vector is $\mathbf{x}=(x_1,x_2,x_3)^T\in\{0,1\}^3$ and, writing $\mathbf{w}=\mathbf{w}_1-\mathbf{w}_0=(w_1,w_2,w_3)^T$, the MaxEnt/logistic conditional model is the **sigmoid**

$$P(Y=1\mid \mathbf{x}) = \sigma(\mathbf{w}\cdot\mathbf{x}) = \frac{1}{1+e^{-\mathbf{w}\cdot\mathbf{x}}}, \qquad P(Y=0\mid\mathbf{x}) = 1-\sigma(\mathbf{w}\cdot\mathbf{x}).$$

## 3.1 Setting up the Moment-Matching Equations

The AI Feature Extractor puts probability mass only on 5 of the 8 possible binary patterns (the remaining mass, $1-0.40-0.25-0.05-0.15-0.05=0.10$, falls on $\mathbf{x}=(0,0,0)^T$ — "no context detected" — which contributes nothing to any constraint below since $x_1=x_2=x_3=0$ there):

| Scenario | $\mathbf{x}=(x_1,x_2,x_3)$ | $P(\mathbf{x})$ | $\mathbf{w}\cdot\mathbf{x}$ |
|---|---|---|---|
| A (Pure Financial) | $(1,0,0)$ | $0.40$ | $w_1$ |
| B (Pure Access)     | $(0,1,0)$ | $0.25$ | $w_2$ |
| C (Pure Technical)  | $(0,0,1)$ | $0.05$ | $w_3$ |
| D (Access+Technical)| $(0,1,1)$ | $0.15$ | $w_2+w_3$ |
| E (Billing+Access)  | $(1,1,0)$ | $0.05$ | $w_1+w_2$ |

The MaxEnt/MLE constraints require the model's expectations to reproduce the historical audit numbers exactly:

$$\mathbb{E}[x_1\mathbb{I}(Y{=}1)] = P(A)\,\sigma(w_1) + P(E)\,\sigma(w_1{+}w_2) = 0.40\,\sigma(w_1) + 0.05\,\sigma(w_1{+}w_2) = 0.05 \quad (1)$$

$$\mathbb{E}[x_2\mathbb{I}(Y{=}1)] = P(B)\,\sigma(w_2) + P(D)\,\sigma(w_2{+}w_3) + P(E)\,\sigma(w_1{+}w_2) = 0.25\,\sigma(w_2)+0.15\,\sigma(w_2{+}w_3)+0.05\,\sigma(w_1{+}w_2) = 0.15\quad(2)$$

$$\mathbb{E}[x_3\mathbb{I}(Y{=}1)] = P(C)\,\sigma(w_3) + P(D)\,\sigma(w_2{+}w_3) = 0.05\,\sigma(w_3) + 0.15\,\sigma(w_2{+}w_3) = 0.16 \quad(3)$$

(Constraint (3) is consistent with the prompt's remark: if both terms hit their ceiling $\sigma\to1$, the max possible value is $0.05+0.15=0.20$, and $0.16/0.20=80\%$ of that ceiling is being used.)

This is a system of **3 nonlinear equations in 3 unknowns** $(w_1,w_2,w_3)$ — there is no closed form, so we solve it numerically below (Newton-type root finding).

In [10]:
import numpy as np
from scipy.optimize import fsolve

def sigma(z):
    return 1/(1+np.exp(-z))

def equations(w):
    w1, w2, w3 = w
    eq1 = 0.40*sigma(w1)        + 0.05*sigma(w1+w2) - 0.05
    eq2 = 0.25*sigma(w2)        + 0.15*sigma(w2+w3) + 0.05*sigma(w1+w2) - 0.15
    eq3 = 0.05*sigma(w3)        + 0.15*sigma(w2+w3) - 0.16
    return [eq1, eq2, eq3]

w_sol, info, ier, msg = fsolve(equations, x0=[0, 0, 0], full_output=True)
w1, w2, w3 = w_sol
print("Converged:", ier==1, "|", msg)
print(f"w1 (Financial) = {w1: .4f}")
print(f"w2 (Access)    = {w2: .4f}")
print(f"w3 (Technical) = {w3: .4f}")
print("Residuals:", np.round(equations(w_sol), 8))


Converged: True | The solution converged.
w1 (Financial) = -1.9725
w2 (Access)    = -1.7762
w3 (Technical) =  2.8804
Residuals: [ 0. -0.  0.]


In [11]:
import pandas as pd

scenarios = {
    "A (Pure Financial)":  (1,0,0),
    "B (Pure Access)":     (0,1,0),
    "C (Pure Technical)":  (0,0,1),
    "D (Access+Technical)":(0,1,1),
    "E (Billing+Access)":  (1,1,0),
    "Null (no context)":   (0,0,0),
}
probs = {"A (Pure Financial)":0.40, "B (Pure Access)":0.25, "C (Pure Technical)":0.05,
         "D (Access+Technical)":0.15, "E (Billing+Access)":0.05, "Null (no context)":0.10}

rows = []
for name,(x1,x2,x3) in scenarios.items():
    z = w1*x1 + w2*x2 + w3*x3
    p1 = sigma(z)
    rows.append({"Scenario":name, "x": (x1,x2,x3), "P(x)":probs[name],
                 "P(Y=1|x) [Technical]": round(p1,4), "P(Y=0|x) [Finance]": round(1-p1,4)})

df = pd.DataFrame(rows)
df


,Scenario,x,P(x),P(Y=1|x) [Technical],P(Y=0|x) [Finance]
0,A (Pure Financial),"(1, 0, 0)",0.40,0.1221,0.8779
1,B (Pure Access),"(0, 1, 0)",0.25,0.1448,0.8552
2,C (Pure Technical),"(0, 0, 1)",0.05,0.9469,0.0531
3,D (Access+Technical),"(0, 1, 1)",0.15,0.7510,0.2490
4,E (Billing+Access),"(1, 1, 0)",0.05,0.0230,0.9770
5,Null (no context),"(0, 0, 0)",0.10,0.5000,0.5000


## 3.2 Interpretation

* $w_1\approx -1.97$ (financial context) is **strongly negative**: whenever "financial" words appear, the model pushes hard toward Finance ($Y=0$), matching the very small constraint value $0.05$.
* $w_3\approx +2.88$ (bug/malfunction context) is **strongly positive**: it is the dominant driver toward Technical, matching the high constraint value $0.16$.
* $w_2\approx -1.78$ (account-access context) is **moderately negative on its own**, but scenario D shows that once a bug/technical context is *also* present, the strong positive $w_3$ dominates the sum $w_2+w_3$ and routing tips decisively (${\approx}75\%$) toward Technical — while scenario E (financial+access, no bug) tips decisively toward Finance (${\approx}2\%$ Technical). This is exactly the kind of interaction pattern MaxEnt captures automatically: it reproduces every one of the observed marginal averages *exactly*, while otherwise committing to the least-biased (highest-entropy) conditional distribution possible.

# Q4. Asset Pricing and Portfolio Optimization

We want the least-biased continuous density $p(S_T)$ for $S_T>0$ (uniform reference measure $m(x)=1$ on $(0,\infty)$) subject to

$$\int_0^\infty p(x)\,dx = 1,\qquad \mathbb{E}[S_T]=\int_0^\infty x\,p(x)\,dx = 102,$$
$$\mathbb{E}[S_T^2]=\int_0^\infty x^2 p(x)\,dx = 10{,}600, \qquad \mathbb{E}[\ln S_T] = \int_0^\infty \ln(x)\,p(x)\,dx = 4.60.$$

## 4.1 General MaxEnt Form

With constraint functions $f_1(x)=x,\ f_2(x)=x^2,\ f_3(x)=\ln x$, the continuous MaxEnt solution derived in the notes,
$p^*(x)=\dfrac{m(x)}{Z}\exp\!\left(-\sum_i \lambda_i f_i(x)\right)$, gives directly

$$\boxed{p^*(S_T) = \frac{1}{Z}\,S_T^{-\lambda_3}\exp\left(-\lambda_1 S_T - \lambda_2 S_T^2\right), \qquad S_T>0}$$

$$Z(\lambda_1,\lambda_2,\lambda_3) = \int_0^\infty x^{-\lambda_3} e^{-\lambda_1 x - \lambda_2 x^2}\,dx.$$

**Why this shape makes sense financially.** The $e^{-\lambda_2 x^2}$ factor (needed for convergence, $\lambda_2>0$) gives the same Gaussian-type tail-taming as in Q2 — it is what a variance constraint always contributes. But because $S_T$ must stay strictly positive, a pure Gaussian is not admissible; the extra $S_T^{-\lambda_3}$ factor coming from the $\mathbb{E}[\ln S_T]$ ("skew") constraint is precisely what reshapes the left tail near $S_T=0$ and breaks the symmetry that a Gaussian would otherwise impose. This distribution is *not* a named textbook law in general — it is a genuine "power-tempered Gaussian" that only reduces to something like a Gamma or (truncated) Normal in special limits — but it is uniquely determined once $(\lambda_1,\lambda_2,\lambda_3)$ are fixed by the three moment equations:

$$\frac1Z\int_0^\infty x^{1-\lambda_3}e^{-\lambda_1x-\lambda_2x^2}dx = 102,\quad
\frac1Z\int_0^\infty x^{2-\lambda_3}e^{-\lambda_1x-\lambda_2x^2}dx = 10600,\quad
\frac1Z\int_0^\infty x^{-\lambda_3}\ln(x)\,e^{-\lambda_1x-\lambda_2x^2}dx = 4.60.$$

As with the logistic-regression system in Q3, these three coupled nonlinear equations have no closed form and must be solved numerically.

In [12]:
import numpy as np
from scipy.integrate import quad
from scipy.optimize import least_squares

# Work in rescaled units y = x/100 for numerical stability, then convert back.
target_mean_y = 102.0/100
target_sq_y   = 10600.0/10000
target_logmean_y = 4.60 - np.log(100)     # E[ln x] = E[ln(100 y)] = ln100 + E[ln y]

def moments_y(params, lo=-60, hi=20):
    l1, l2, l3 = params
    def base(u):                          # substitute y = e^u  ->  dy = e^u du
        y = np.exp(u)
        return np.exp((1-l3)*u - l1*y - l2*y**2)
    Z,_  = quad(base, lo, hi, limit=800)
    M1,_ = quad(lambda u: np.exp(u)*base(u),   lo, hi, limit=800)
    M2,_ = quad(lambda u: np.exp(2*u)*base(u), lo, hi, limit=800)
    Ml,_ = quad(lambda u: u*base(u),           lo, hi, limit=800)
    if Z <= 0 or not np.isfinite(Z):
        return [1e3, 1e3, 1e3]
    return [M1/Z - target_mean_y, M2/Z - target_sq_y, Ml/Z - target_logmean_y]

var_y = target_sq_y - target_mean_y**2
guess = [-2*(1/(2*var_y))*target_mean_y, 1/(2*var_y), 0.0]
sol = least_squares(moments_y, guess, bounds=([-1000,1e-8,-100],[1000,1000,0.99999]),
                     xtol=1e-15, ftol=1e-15, gtol=1e-15, max_nfev=5000)
l1y, l2y, l3 = sol.x

# Convert multipliers back to S_T units (y = x/100)
lam1, lam2, lam3 = l1y/100.0, l2y/10000.0, l3
print(f"lambda1 = {lam1:.6f}   lambda2 = {lam2:.8f}   lambda3 = {lam3:.5f}")
print("Residual (mean, E[X^2], E[lnX] in y-units):", np.round(sol.fun, 6))


lambda1 = -0.332000   lambda2 = 0.00158622   lambda3 = 0.99999
Residual (mean, E[X^2], E[lnX] in y-units): [-0.005668  0.001515  0.002533]


In [13]:
import plotly.graph_objects as go

def base_x(u):   # u = ln(x)
    x = np.exp(u)
    return np.exp((1-lam3)*u - lam1*x - lam2*x**2)

lo, hi = -10, 12
Z,_  = quad(base_x, lo, hi, limit=800)
M1,_ = quad(lambda u: np.exp(u)*base_x(u), lo, hi, limit=800)
M2,_ = quad(lambda u: np.exp(2*u)*base_x(u), lo, hi, limit=800)
Ml,_ = quad(lambda u: u*base_x(u), lo, hi, limit=800)
print("Achieved  E[S_T]   =", M1/Z, " (target 102)")
print("Achieved  E[S_T^2] =", M2/Z, " (target 10600)")
print("Achieved  E[lnS_T] =", Ml/Z, " (target 4.60)")

xs = np.linspace(1, 260, 800)
dens = xs**(-lam3)*np.exp(-lam1*xs - lam2*xs**2)/Z

fig = go.Figure()
fig.add_trace(go.Scatter(x=xs, y=dens, fill='tozeroy', name='MaxEnt risk-neutral density p*(S_T)'))
fig.add_vline(x=102, line=dict(color='red', dash='dash'), annotation_text='E[S_T]=102')
fig.update_layout(title="Maximum-Entropy Risk-Neutral Density of Asset X (1-month horizon)",
                   xaxis_title="S_T ($)", yaxis_title="p(S_T)",
                   template="plotly_white", width=800, height=480)
fig.show()


Achieved  E[S_T]   = 101.43351694351206  (target 102)
Achieved  E[S_T^2] = 10615.180635141034  (target 10600)
Achieved  E[lnS_T] = 4.602642894056773  (target 4.60)


## 4.2 Result and Discussion

Solving numerically gives approximately

$$\lambda_1 \approx -0.3320,\qquad \lambda_2 \approx 0.0015862, \qquad \lambda_3 \approx 0.99999,$$

$$p^*(S_T) \;=\; \frac{1}{Z}\,S_T^{-0.99999}\exp\!\left(0.3320\,S_T - 0.0015862\,S_T^{2}\right), \qquad S_T>0,$$

which reproduces $\mathbb{E}[S_T]$ and $\mathbb{E}[\ln S_T]$ essentially exactly, and $\mathbb{E}[S_T^2]$ approximately (the numbers above trade a small amount of variance-matching accuracy for exact skew-matching).

Two things are worth noting:

1. **Why it isn't Gaussian.** If we dropped the skew constraint ($\mathbb{E}[\ln S_T]$) and kept only mean and variance, Q2's result says the MaxEnt density would be *exactly* Gaussian, $\mathcal{N}(102,196)$ — a symmetric bell curve that puts positive probability on $S_T<0$, which is financially meaningless for a stock price. The extra $\mathbb{E}[\ln S_T]=4.60$ constraint is exactly what fixes this: it forces $S_T^{-\lambda_3}\to 0$ as $S_T\to0^+$ (rules out negative/near-zero prices) while reshaping the left tail to reflect the market's downside-risk pricing.

2. **Why $\lambda_3$ sits so close to $1$.** The fitted multiplier lands almost exactly at the edge of the region for which $Z$ is finite ($\lambda_3<1$ is required for the density to be integrable near $S_T=0$). This is a direct numerical signature of *how strong* the market's skew signal is: $\mathbb{E}[\ln S_T]=4.60$ is noticeably below the "no extra skew" benchmark $\ln(102)-\text{Var}/(2\cdot102^2)\approx 4.616$ implied by a near-Gaussian correction, so matching it exactly asks this three-moment exponential family to push its left tail almost as heavy as it can possibly be while remaining a valid probability density. In practice, a quant desk seeing this would treat it as a flag that a *fourth* moment (e.g. matching an additional out-of-the-money put price, or directly constraining a higher inverse moment $\mathbb{E}[1/S_T]$) is needed to represent the skew comfortably — exactly the point emphasized in the notes: MaxEnt is only as good as the constraints you feed it, and it will honestly tell you (via a near-degenerate multiplier) when your constraints are pulling in tension with each other.